docs: env api expand runtime agnostic SSR section (#18334)
vitejs/vite: docs/guide/api-environment-frameworks.md

## Environments and frameworks

> The implicit `ssr` environment and other non-client environments use a `RunnableDevEnvironment` by default during dev. While this requires the runtime to be the same with the one the Vite server is running in, this works similarly with `ssrLoadModule` and allows frameworks for to migrate and enable HMR during for their SSR dev story. You can guard any runnable environment with an `isRunnableDevEnvironment` function.

Reviewer: ```suggestion
The implicit `ssr` environment and other non-client environments use a `RunnableDevEnvironment` by default during dev. While this requires the runtime to be the same with the one the Vite server is running in, this works similarly with `ssrLoadModule` and allows frameworks to migrate and enable HMR for their SSR dev story. You can guard any runnable environment with an `isRunnableDevEnvironment` function.
```

:::

`RunnableDevEnvironment` has a `runner.import` function that returns the value of the module. But this function is not available in the raw `DevEnvironment` and requires the code using the Vite's APIs and the user modules to be decoupled.

For example, the following example uses the value of the user module from the code using the Vite's APIs:

```ts
// code using the Vite's APIs
import { createServer } from 'vite'

const server = createServer()
const ssrEnvironment = server.environment.ssr
const input = {}

const { createHandler } = await ssrEnvironment.runner.import('./entrypoint.js')
const handler = createHandler(input)
const response = handler(new Request('/'))

// -------------------------------------
// ./entrypoint.js
export function createHandler(input) {
  return function handler(req) {
    return new Response('hello')
  }
}
```

> If your code can run in the same runtime with the user modules (= does not rely on Node.js specific APIs), you can use a virtual module so that you don't need to access the value from the code using the Vite's APIs.

Reviewer 2: ```suggestion
If your code can run in the same runtime as the user modules (i.e., it does not rely on Node.js-specific APIs), you can use a virtual module. This approach eliminates the need to access the value from the code using Vite's APIs.
```

```ts
// code using the Vite's APIs
import { createServer } from 'vite'

const server = createServer({
  plugins: [
    // a plugin that handles `virtual:entrypoint`
    {
      name: 'virtual-module',
      /* plugin implementation */
    },
  ],
})
const ssrEnvironment = server.environment.ssr
const input = {}

// use exposed functions by each environment factories that runs the code
// check for each environment factories what they provide
if (ssrEnvironment instanceof RunnableDevEnvironment) {
  ssrEnvironment.runner.import('virtual:entrypoint')
} else if (ssrEnvironment instanceof CloudflareDevEnvironment) {

>   ssrEnvironment.evaluate('virtual:entrypoint')

Reviewer 2: maybe it's better to have a fetchable API here? The `evaluate` and `cloudflare` are a bit confusing here - especially because `cloudflare` is a fetchable environment

Author: I've pushed a commit that replaces them with `CustomDevEnvironment` and `.runEntrypoint` so that it doesn't look like it can satisfy the `FetchableDevEnvironment` interface.

}
```

> If your code requires Node.js APIs, you can use `hot.send` to communicate with the code that uses Vite's APIs from the user modules. Note that you might not achieve the same thing after build.

Reviewer 2: ```suggestion
If your code requires Node.js APIs, you can use `hot.send` to communicate with the code that uses Vite's APIs from the user modules. However, be aware that this approach may not work the same way after the build process.
```
